load-gen/src/main/java/io/smartcat/data/loader/datasource/BufferedDataSource.java

```diff
@@ -0,0 +1,99 @@
+ *
+ * @param <T> Type of the data provided by data source implementation.
+ */
+public class BufferedDataSource<T> implements DataSource<T> {
+
+    private final Object bufferFilled = new Object();
+    private final DataSource<T> delegate;
+    private final int bufferSize;
+    private final Queue<T> buffer;
+    private final double initialBufferFullness;
+
+    private AtomicBoolean terminateThread = new AtomicBoolean(false);
+    private Thread thread;
+
+    /**
+     * Constructs a buffered data source with specified <code>delegate</code> data source. <code>bufferSize</code> is
+     * set to 1024, <code>initialBufferFullness</code> is set to 75%. Blocks polling at the beginning until buffer is
+     * filled to <code>initialBufferFullness</code>.
+     *
+     * @param delegate Data source which will be used to fill buffer.
+     */
+    public BufferedDataSource(DataSource<T> delegate) {
+        this(delegate, 1024);
+    }
+
+    /**
+     * Constructs a buffered data source with specified <code>delegate</code> data source and <code>bufferSize</code>.
+     * <code>initialBufferFullness</code> is set to 75%. Blocks polling at the beginning until buffer is filled to
+     * <code>initialBufferFullness</code>.
+     *
+     * @param delegate Data source which will be used to fill buffer.
+     * @param bufferSize Size of the buffer.
+     */
+    public BufferedDataSource(DataSource<T> delegate, int bufferSize) {
+        this(delegate, bufferSize, 0.75);
+    }
+
+    /**
+     * Constructs a buffered data source with specified <code>delegate</code> data source, <code>bufferSize</code> and
+     * <code>initialBufferFullness</code>. Blocks polling at the beginning until buffer is filled to
+     * <code>initialBufferFullness</code>.
+     *
+     * @param delegate Data source which will be used to fill buffer.
+     * @param bufferSize Size of the buffer.
+     * @param initialBufferFullness Initial fullness of buffer.
+     */
+    public BufferedDataSource(DataSource<T> delegate, int bufferSize, double initialBufferFullness) {
+        this.delegate = delegate;
+        this.bufferSize = bufferSize;
+        this.buffer = new LinkedBlockingQueue<>(bufferSize);
+        this.initialBufferFullness = initialBufferFullness;
+        fillBuffer();
+    }
+
+    @Override
+    public boolean hasNext(long time) {
+        return !buffer.isEmpty() || delegate.hasNext(time);
+    }
+
+    @Override
+    public T getNext(long time) {
+        try {
+            bufferFilled.wait();
+            return buffer.poll();
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void terminate() {
+        terminateThread.set(false);
+        delegate.terminate();
+    }
+
+    private void fillBuffer() {
+        thread = new Thread(() -> {
+            while (buffer.size() < bufferSize * initialBufferFullness) {
+                buffer.offer(delegate.getNext(0));
```

Comment:
Yes, it breaks. Do you have idea how can we/someone implement buffer if needed and not break the contract? Problem is here because  buffering is done from another thread where we don't have info on time.